### PR TITLE
Item 6425: Changes to support user profile management within apps

### PR DESCRIPTION
**Pull request by cnathe**

### CODE REVIEW

**Comment by cnathe** [comment]
- Move validation code from ShowUpdateAction to UsersTableQueryUpdateService 
- add UserController.UpdateUserDetailsAction, api version of ShowUpdateAction
- add LoginController.GetPasswordRulesInfoAction
- add LoginController.ChangePasswordApiAction, refactor current ChangePasswordAction to share validation code

File: api/src/org/labkey/api/query/QueryUpdateForm.java [hunk 1]
```diff
@@ -35,12 +35,19 @@
      * Prefix prepended to all form elements
      */
     public static final String PREFIX = "quf_";
+    private boolean _ignorePrefix;
```
[the remark this answers is not in the document]
Comment (cnathe) [reply api/src/org/labkey/api/query/QueryUpdateForm.java]: Done.

File: core/src/org/labkey/core/query/UsersTable.java [hunk 2]
```diff
@@ -419,4 +444,191 @@ public void fireRowTrigger(Container c, User user, TriggerType type, boolean bef
...
+            validateExpirationDate(userToUpdate, user, container, row);
+
+            SpringAttachmentFile avatarFile = (SpringAttachmentFile)row.get(UserAvatarDisplayColumnFactory.FIELD_KEY);
+            validateAvatarFile(avatarFile);
+
+            Map<String, Object> ret = super.updateRow(user, container, row, oldRow);
+            updateAvatarFile(userToUpdate, avatarFile, row);
+
+            if (row.containsKey(EXPIRATION_DATE_KEY))
+                auditExpirationDateChange(userToUpdate, user, ContainerManager.getRoot(), userToUpdate.getExpirationDate(), (Date)ret.get(EXPIRATION_DATE_KEY));
```
[the remark this answers is not in the document]
Comment (cnathe) [reply core/src/org/labkey/core/query/UsersTable.java]: No, it does not look like we do. This code was taken from the current ShowUpdateAction for the LKS user profile page and moved here to the UsersTable QUS, so no functional changes. Looks like we audit password set/change, expiration date change, but not user detail changes.

File: core/src/org/labkey/core/query/UsersTable.java [hunk 3]
```diff
@@ -419,4 +444,191 @@ public void fireRowTrigger(Container c, User user, TriggerType type, boolean bef
...
+        private void validatePermissions(User editingUser, User userToUpdate)
+        {
+            // only allow update to your own record or if you are a site level user manager
+            boolean isOwnRecord = editingUser.equals(userToUpdate);
+            if (!editingUser.hasRootPermission(UserManagementPermission.class) && !isOwnRecord)
+                throw new UnauthorizedException();
+
+            // don't let non-site admin edit details of site admin account
+            if (userToUpdate.hasSiteAdminPermission() && !editingUser.hasSiteAdminPermission())
+                throw new UnauthorizedException("Can not edit details for a Site Admin user.");
```
[the remark this answers is not in the document]
Comment (cnathe) [reply core/src/org/labkey/core/query/UsersTable.java]: Done.

File: core/src/org/labkey/core/query/UsersTable.java [hunk 4]
```diff
@@ -419,4 +444,191 @@ public void fireRowTrigger(Container c, User user, TriggerType type, boolean bef
...
+                }
+            }
+        }
+
+        private void updateAvatarFile(User user, SpringAttachmentFile file, Map<String, Object> row) throws ValidationException
+        {
+            ThumbnailService.ImageType imageType = ThumbnailService.ImageType.Large;
+            ThumbnailService svc = ThumbnailService.get();
+
+            if (svc != null)
```
[the remark this answers is not in the document]
Comment (cnathe) [reply core/src/org/labkey/core/query/UsersTable.java]: Sure. I agree that I don't know how svc could be null, but I will add an error.

File: core/src/org/labkey/core/user/UserController.java [hunk 5]
```diff
@@ -1068,115 +976,85 @@ public boolean handlePost(QueryUpdateForm form, BindException errors) throws Exc
...
+                List<Map<String, Object>> keys = new ArrayList<>(Arrays.asList(Map.of("UserId", form.getUserId())));
 
-        @Override
-        public NavTree appendNavTrail(NavTree root)
-        {
-            addUserDetailsNavTrail(root, _pkVal);
-            root.addChild("Update");
-            return root.addChild(UserManager.getEmailForId(_pkVal));
+                response.put("updatedRows", table.getUpdateService().updateRows(getUser(), root, rows, keys, null, null));
+                response.put("success", true);
```
[the remark this answers is not in the document]
Comment (cnathe) [reply core/src/org/labkey/core/user/UserController.java]: Yes, it if reaches this line, then the QUS.updateRows was successful and didn't throw any errors/exceptions.

File: core/src/org/labkey/core/query/UsersTable.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (labkey-susanh) [reply core/src/org/labkey/core/query/UsersTable.java]: I guess this is as it should be.

**Pull request merged**